modules/apps/portal-workflow/portal-workflow-kaleo-runtime-impl/src/main/java/com/liferay/portal/workflow/kaleo/runtime/internal/action/executor/ActionExecutorManagerImpl.java

```diff
@@ -114,6 +131,18 @@ private String _getActionExecutorKey(KaleoAction kaleoAction) {
 		return kaleoAction.getScriptLanguage();
 	}
 
+	private List<ActionExecutor> _getActionExecutors(String actionExecutorKey) {
+		List<ActionExecutor> actionExecutors = _serviceTrackerMap.getService(
+			_encodeKey(actionExecutorKey, CompanyConstants.SYSTEM));
+
+		if (actionExecutors != null) {
+			return actionExecutors;
+		}
+
+		return _serviceTrackerMap.getService(
```

Comment:
Shouldn't the company instance ones be merged with the SYSTEM ones?

I believe the SYSTEM ones should be available to all instances.

Reply:
@rotty3000, with the suggested logic, the System ones are available to all instances

Reply:
@rotty3000 I want to clarify something, we are not adding support to deploy Client Extensions to be used in all instances by setting 0 as company. We just use the System company as way to generate a key for those global cases where the definition of a company scope does not make sense (for example _UpdateStatusActionExecutor_)

Reply:
Let me ask a question.

Does this method not return the list that results in the options that will appear in the UI drop down?

If so, that list should contain both the SYSTEM functions as well as the ones scoped to the current company instance. Otherwise, how can someone in a virtual instance use a SYSTEM function in their definition?

Reply:
I think we were talking about different things. This method does not return what we see in the UI drop down list. It returns the list of actionExecutors associated to a specific _actionExecutorKey_ (to be honest I do not how one key can have several actionExecutors) just before executing it.
This is the call: https://github.com/liferay-workflow/liferay-portal/pull/1117/files/8a30d6e4a0cbcea3b09ed8be912268892478f256#diff-98485ca8f3b52bb5926c8678e04b0b6e5a02a712d0c07f2236f8fdf5894a53a6R49

And we have two kind of actionExecutors:
- System ones (not tied to any company) like updateStatus, Groovy, etc.
- Company scoped functions like the ones we create via CX

So, at least at this moment, there is no way that an actionExecutor can be tied to both, system and company scopes.

The method that is used to populate the UI drop down is this one:
https://github.com/liferay-workflow/liferay-portal/pull/1117/files/8a30d6e4a0cbcea3b09ed8be912268892478f256#diff-98485ca8f3b52bb5926c8678e04b0b6e5a02a712d0c07f2236f8fdf5894a53a6L79

I hope the logic is more clear now